### PR TITLE
PLAT-101792: Fix not to scroll abnormally while pressing 5-way keys after wheeling

### PR DESCRIPTION
**Pull request by ybsung**
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [x] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [ ] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

1. After wheeling in native VirtualList,
2. if pressing 5-way keys,
3. then the list suddenly scrolls.
4. So the spotted item is stick to the viewport.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

While wheeling in native VirtualList, the final scroll position was not cached properly. So if pressing 5-way keys on the list item, then the wrong scroll position was referenced. Therefore for native VirtualList, the real scroll position instead of the cached scroll position is used.

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

### Links
[//]: # (Related issues, references)

PLAT-101792

### CODE REVIEW

File: packages/ui/Scroller/ScrollerBasic.js [hunk 1]
```diff
@@ -115,7 +115,30 @@ class ScrollerBasic extends Component {
...
+	}
+
+	// scrollMode 'native' only
+	didScroll (x, y) {
+		this.scrollPos.left = x;
+		this.scrollPos.top = y;
+	}
+
+	// scrollMode 'translate' only
+	setScrollPositionTarget (x, y) {
```
Comment (MikyungKim) [inline packages/ui/Scroller/ScrollerBasic.js]: I don't think we need this function in `Scroller`.
There is no reference case for `scrollPositionTarget` from themed `Scroller`.
Why did you add this?

File: packages/ui/VirtualList/VirtualListBasic.js [hunk 2]
```diff
@@ -470,6 +470,7 @@ class VirtualListBasic extends Component {
 		if (this.props.scrollMode === 'native') {
 			this.scrollToPosition(x, y, rtl);
 		} else {
+			this.setScrollPositionTarget (x, y);
 			this.setScrollPosition(x, y, rtl, x, y);
```
Comment (MikyungKim) [inline packages/ui/VirtualList/VirtualListBasic.js]: You are passing additional parameter `x`, `y` here.

File: packages/ui/Scroller/ScrollerBasic.js (hunk 1 above)
Comment (ybsung) [reply packages/ui/Scroller/ScrollerBasic.js]: Oh, my mistake. I should be removed.

File: packages/ui/CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,12 @@
 
 The following is a curated list of changes in the Enact ui module, newest changes on the top.
 
+## [unreleased]
+
+### Fixed
+
+- `ui/Scroller.Scroller`, `ui/VirtualList.VirtualList` and `ui/VirtualList.VirtualGridList` not to suddenly jump when pressing directional keys after wheeling.
```
Comment (MikyungKim) [inline packages/ui/CHANGELOG.md]: Please omit Scroller here.

File: packages/ui/VirtualList/VirtualListBasic.js [hunk 4]
```diff
@@ -829,19 +824,21 @@ class VirtualListBasic extends Component {
...
 	// scrollMode 'translate' only
-	setScrollPosition (x, y, rtl = this.props.rtl, targetX = 0, targetY = 0) {
+	setScrollPositionTarget (x, y) {
+		// The `x`, `y` as parameters in scrollToPosition() are the position when stopping scrolling.
+		// But the `x`, `y` as parameters in setScrollPosition() are the position between current position and the position stopping scrolling.
+		// To know the position when stopping scrolling properly, `x` and `y` are passed and cached in `this.scrollPositionTarget`.
+		if (this.isPrimaryDirectionVertical) {
+			this.scrollPositionTarget = y;
+		} else {
+			this.scrollPositionTarget = x;
```
Comment (jeonghee27) [inline packages/ui/VirtualList/VirtualListBasic.js]: `setScrollPositionTarget` `setScrollPosition` both function are for 'translate'.
I wonder the reason why you divide it to two function?

Comment (ybsung) [reply packages/ui/VirtualList/VirtualListBasic.js]: `setScrollPositionTarget` is needed to set the position that a scroller will stop.
` setScrollPosition` is needed to update the scroll position while animating and going to the position that a scroller will stop.

To handle 5-way keys in sandstone VirtualList onAcceleratedKeyDown() properly, we need to know the position that a scroller will stop to make a decision if scrolling into a view or not.

File: packages/ui/useScroll/useScroll.js [hunk 5]
```diff
@@ -1076,11 +1076,14 @@ const useScrollBase = (props) => {
...
 			showThumb(bounds);
+			if (scrollContentHandle.current && scrollContentHandle.current.setScrollPositionTarget) {
+				scrollContentHandle.current.setScrollPositionTarget(targetX, targetY);
+			}
 
 			if (animate) {
 				mutableRef.current.animator.animate(scrollAnimation(mutableRef.current.animationInfo));
 			} else {
-				scroll(targetX, targetY, targetX, targetY);
+				scroll(targetX, targetY);
```
Comment (jeonghee27) [inline packages/ui/useScroll/useScroll.js]: Now  `scroll` function get two parameter always.
I think the third parameter make confusing to understand further behavior. 
`scroll = (left, top, ...rest) => {`
How about remove the `...rest` also?

File: packages/ui/VirtualList/VirtualListBasic.js (hunk 4 above)
Comment (jeonghee27) [reply packages/ui/VirtualList/VirtualListBasic.js]: I understand. You wanted to omit update the `scrollPositionTarget` in situations such as animations. Please let me know If I misunderstand,

File: packages/ui/useScroll/useScroll.js (hunk 5 above)
Comment (ybsung) [reply packages/ui/useScroll/useScroll.js]: For `translate` VirtualList, the `rest` should be passed to the `setScrollPosition`.

Comment (ybsung) [reply packages/ui/useScroll/useScroll.js]: So the `this.scroll` usage is
```
this.scroll(targetX, targetY, targetX, targetY);
```

Comment (MikyungKim) [reply packages/ui/useScroll/useScroll.js]: ? I think @jeonghee27 is right.

Comment (MikyungKim) [reply packages/ui/useScroll/useScroll.js]: Now we are setting `scrollPositionTarget` seperately.

Comment (ybsung) [reply packages/ui/useScroll/useScroll.js]: Oh, my misunderstanding. I fixed it.
